vector/read.go

```diff
@@ -14,570 +16,597 @@ import (
 	if demand.IsNone(demandOut) {
 		return &constants{}, nil
 	}
 
-	switch reader := reader.(type) {
+	switch meta := meta.(type) {
 
-	case *vngvector.ArrayReader:
-		lengths, err := readInt64s(reader.Lengths)
+	case *vngvector.Array:
+		lengths, err := ReadInt64s(readerAt, buf, meta.Lengths)
 		if err != nil {
 			return nil, err
 		}
-		elems, err := read(context, reader.Elems, demand.All())
+		elems, err := read(zctx, readerAt, buf, meta.Values, demand.All())
 		if err != nil {
 			return nil, err
 		}
-		vector := &arrays{
+		return &arrays{
 			lengths: lengths,
 			elems:   elems,
-		}
-		return vector, nil
-
-	case *vngvector.ConstReader:
-		var builder zcode.Builder
-		err := reader.Read(&builder)
-		if err != nil {
-			return nil, err
-		}
-		it := zcode.Bytes(builder.Bytes()).Iter()
-		vector := &constants{
-			bytes: it.Next(),
-		}
-		return vector, nil
+		}, nil
 
-	case *vngvector.DictReader:
-		// TODO Would we be better off with a dicts vector?
-		return readPrimitive(context, reader.Typ, func() ([]byte, error) { return reader.ReadBytes() })
+	case *vngvector.Const:
+		return &constants{
+			bytes: meta.Value.Bytes(),
+		}, nil
 
-	case *vngvector.MapReader:
-		keys, err := read(context, reader.Keys, demand.All())
+	case *vngvector.Map:
+		keys, err := read(zctx, readerAt, buf, meta.Keys, demand.All())
 		if err != nil {
 			return nil, err
 		}
-		lengths, err := readInt64s(reader.Lengths)
+		lengths, err := ReadInt64s(readerAt, buf, meta.Lengths)
 		if err != nil {
 			return nil, err
 		}
-		values, err := read(context, reader.Values, demand.All())
+		values, err := read(zctx, readerAt, buf, meta.Values, demand.All())
 		if err != nil {
 			return nil, err
 		}
-		vector := &maps{
+		return &maps{
 			lengths: lengths,
 			keys:    keys,
 			values:  values,
-		}
-		return vector, nil
+		}, nil
 
-	case *vngvector.NullsReader:
-		runs, err := readInt64s(&reader.Runs)
+	case *vngvector.Named:
+		return read(zctx, readerAt, buf, meta.Values, demandOut)
+
+	case *vngvector.Nulls:
+		runs, err := ReadInt64s(readerAt, buf, meta.Runs)
 		if err != nil {
 			return nil, err
 		}
-		values, err := read(context, reader.Values, demandOut)
+		values, err := read(zctx, readerAt, buf, meta.Values, demandOut)
 		if err != nil {
 			return nil, err
 		}
 		if len(runs) == 0 {
 			return values, nil
 		}
-		vector := &nulls{
+		return &nulls{
 			runs:   runs,
 			values: values,
-		}
-		return vector, nil
+		}, nil
 
-	case *vngvector.PrimitiveReader:
-		return readPrimitive(context, reader.Typ, func() ([]byte, error) { return reader.ReadBytes() })
+	case *vngvector.Primitive:
+		if len(meta.Dict) != 0 {
+			tags, err := readSegmap(readerAt, meta.Segmap)
+			if err != nil {
+				return nil, err
+			}
+			return &dict{
+				dict: meta.Dict,
+				tags: tags,
+			}, nil
+		} else {
+			return readPrimitive(zctx, readerAt, buf, meta.Segmap, meta.Type(zctx))
+		}
```

Comment:
Nits: We usually minimize indentation when doing so doesn't otherwise make things less readable. And if an `if` branch block returns, we usually omit `else`.
```suggestion
		if len(meta.Dict) == 0 {
			return readPrimitive(zctx, readerAt, buf, meta.Segmap, meta.Type(zctx))
		}
		tags, err := readSegmap(readerAt, meta.Segmap)
		if err != nil {
			return nil, err
		}
		return &dict{
			dict: meta.Dict,
			tags: tags,
		}, nil
```